lib/ecto/changeset.ex

```diff
@@ -1193,6 +1219,11 @@ defmodule Ecto.Changeset do
     %{changeset | changes: changes, errors: errors, valid?: valid?}
   end
 
+  defp put_change(data, changes, errors, valid?, key, value, {:parameterized, Ecto.Type.Embed, opts}) do
+    relation = Ecto.Type.Embed.get_relation(opts)
+    put_change(data, changes, errors, valid?,key, value, {:embed, relation})
```

Comment:
This should probably be a new callback, right?

Reply:
Yes, part of the issue was that a lot of my iterations on this went down rabbit holes of supporting the existing put_* and cast_* relation functions, and I had tried it on a number of different levels inside the existing logic. This being my first deep dive into how Ecto works under the hood, there was a lot of TDD where I was simply trying to clean up the things that broke when it saw a :parameterized tuple. I finally stepped back from that so I could concentrate on the direct `cast`, `load`, and `dump` paths, which allowed me to make progress. I also shifted towards doing as much as possible in the Embed function itself, and not trying to "fix" the logic inside changeset surrounding the "relation" object. I think this is an instance of me finding a path that worked early on (by faking a relation object), and I need to go back and figure out if/when I should be jumping out to a callback in parameterzed type as opposed to simply massaging the data and hooking into the current put_change logic.